[SYCL][Spec] Update the matrix spec based on new use argument

dkhaldi opened the pull request.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,651 @@
+
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+  template <typename Group, typename Ta, typename Tb, typename Tc,
````
Does Intel hardware support `Ta != Tb`? I'm surprised if they would ever be different types?

Also in the CUDA + HIP AMD backends it is not necessary to provide the matrix_layout (doc should be updated to "layout" at some point) template parameter to `joint_matrix_mad`: it looks like it is always necessary for intel hardware? if so I can add it to the CUDA implementation so that the interfaces match.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,651 @@
+  template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          matrix_use U, matrix_layout L,
+          access::address_space Space>	  
+  void joint_matrix_store(Group sg, joint_matrix<T, NumRows, NumCols, U, L, Group> &res,
```
Ah I forgot that Intel hardware supports store for non accumulator matrices, I'll update https://github.com/intel/llvm/pull/6657 adding the use template parameter and throw a runtime error if it is not `use::accumulator` in the CUDA backend.

dkhaldi replied:
We remove L from the template arguments of store and replace it with "unused" in joint matrix.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,651 @@
+
+Note that for getting maximum performance on Intel AMX and DPAS, prepacking data in the memory is necessary. If users did not specify the packed layouts, transforms done by the implementation will be slow due to extra scatter/gather operations. Hence, we expose the `packed` layout to the user to specify that A or B have already been VNNIed. The packed or VNNI layout is introduced in `VNNI layout` ...
+	
+IMPORTANT: In the current AMX and DPAS implementation, the layout in the load of matrix B must be `packed_b` or `row_major`. Automatic VNNI transform is supported. The layout in the load of matrices A and C must be `row_major`, and the layout in the store of matrix C must also be `row_major`.
```
```suggestion
IMPORTANT: In the current AMX and DPAS implementation, the layout in the load of matrix B must be `packed` or `row_major`. Automatic VNNI transform is supported. The layout in the load of matrices A and C must be `row_major`, and the layout in the store of matrix C must also be `row_major`.
```

I guess if you are removing `packed_b` and `packed_a` this sentence will change like this?

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,651 @@
+
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+template <typename T, size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent, 
````
There is a potential issue here: Rows and Cols have default values, but there isn't a default value for the `use` parameter. Assuming the Rows/Cols default value is necessary long term (it isn't in the CUDA backend), should we switch the order of the parameters here?

dkhaldi replied:
@yubingex007-a11y, "use" implementation seems to override the default values for rows and cols. Can you please check that?
I think use should come second in order so we don't have to set a default value for it

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,651 @@
+          matrix_use U, matrix_layout L,
+          access::address_space Space>
+  void joint_matrix_load(Group sg, joint_matrix<T, NumRows, NumCols, U, L, Group> &res,
+		    multi_ptr<T, Space> src, size_t stride, matrix_layout memL);
```
Can you clarify the distinction between the `matrix_layout L` template parameter and `matrix_layout memL` parameter here?

JackAKirk replied:
Specifically do you plan on keeping the `matrix_layout L` template parameter in `joint_matrix_load` and `joint_matrix_store`?
Perhaps you have only kept this in because you think it is necessary for compatibility with the CUDA backend, since I notice that in the latest implementation, https://github.com/intel/llvm/pull/5835, L is set as layout::unused by default?
In the CUDA implementation I took the approach of overloading `joint_matrix_load`: the first overload, https://github.com/intel/llvm/blob/e55e5f0bfe5175d41f61f4b912243a988104e2f7/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L752, was intended to be compatible with the Intel backend, because `layout` is provided as a normal runtime parameter. This is used in the CUDA backend to load the accumulator matrices.
The second overload, https://github.com/intel/llvm/blob/e55e5f0bfe5175d41f61f4b912243a988104e2f7/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp#L780, was intended to only be used in the CUDA and HIP AMD backends, because it is required for the A/B "multiplicand" matrices for which the layout is required at compile time. In this overload there is no runtime `layout` parameter; instead the layouts of the matrices are inferred from the `joint_matrix` arguments.

In the Intel backend can the layouts of all matrices be specified at runtime? Or is it the case that, as with the CUDA backend, some of them require that the layout is known at compile time?

dkhaldi replied:
Same here, we remove L from the template arguments of store and replace it with "unused" in joint matrix.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,651 @@
+          std::size_t M, std::size_t K, std::size_t N,
+	  matrix_layout La, matrix_layout Lb,
+          matrix_layout Lc>
+  joint_matrix<Group, Tc, M, N, matrix_use::accumulator, Lc, Group> joint_matrix_mad(Group sg, joint_matrix<Ta, M, K, matrix_use::matrix_a, La, Group> A,
```
What happens if we just add a new type for D matrix but keep it as a returned argument of joint_matrix_mad?

JackAKirk replied:
The issue is that the data type held by the D matrix can differ from the data type held by the C matrix. For example if type C is fp16 and type D is fp32; does the intel backend not have such cases? The only way to deal with this is using information provided by the user via arguments to joint_matrix_mad. The cleanest way to do this is the above change where the data type of D can be inferred from the joint_matrix D argument. This change allows us to implement two missing mma operations in the CUDA backend that have been added in this commit: https://github.com/intel/llvm/pull/6657/commits/e55e5f0bfe5175d41f61f4b912243a988104e2f7.

 Another option is if there is an optional "output precision argument" that will specify the type of D: in any case I don't see how a fourth argument to joint_matrix_mad can be avoided if we want to enable the missing cases introduced in https://github.com/intel/llvm/pull/6657/commits/e55e5f0bfe5175d41f61f4b912243a988104e2f7

The only other option I see is to not allow the type of D to differ from the type of C. I don't see how adding a new type for D matrix and keeping it as a returned argument can affect this in any way.

dkhaldi replied:
We agreed the first resolution here is to add a new template type Td.
Also, replace Lc, La, Lb by unused

dkhaldi replied:
I thought for the CUDA backend, LA and LB are needed so we should replace them by unused here, right?

JackAKirk replied:
> @JackAKirk, I thought for the CUDA backend, LA and LB are needed so we should replace them by unused here, right?

Yes that's right.

dkhaldi replied:
I meant we should NOT replace them, sorry.

JackAKirk replied:
Yeah I meant to keep the template parameters but set them as default unused since I think that in the Intel backend you don't use them, and always use joint_matrix with layout::unused? In the CUDA backend they can be inferred from the joint_matrix so also don't have to be explicitly specified.

dkhaldi replied:
That's right, good

JackAKirk commented:
By the way, I corrected a lot of typos that are still relevant for the new document here: https://github.com/intel/llvm/pull/6525/files

Basically if you ignore the new sections I added there (#### Binary Multiply and Add and ##### `wi_data` as an `marray` for Nvidia® Tensor Cores) everything else is just small corrections that you could add here.

dkhaldi commented:
> By the way, I corrected a lot of typos that are still relevant for the new document here: https://github.com/intel/llvm/pull/6525/files
> 
> Basically if you ignore the new sections I added there (#### Binary Multiply and Add and ##### `wi_data` as an `marray` for Nvidia® Tensor Cores) everything else is just small corrections that you could consider adding here.

Yes, I will add these corrections here. Thanks.

gmlueck commented:
I have a question about the old specification, which is now named "sycl_ext_oneapi_deprecated_matrix_no_use".  Is that version of the API still supported?  Does (or will) the implementation print a warning message for applications that use the old API?

Or, is the old API no longer supported, and it is (or will be) removed?

Both options are OK for an experimental extension like this.  However, we should treat the documentation differently in the  two cases.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+IMPORTANT: In the current implementation, only the static extent is supported
+
+#### Use
+Specifying the usage of the matrix: matrix left (A), matrix right (B) or accumuator (C) is required by backend implementations to reason about the layout of the matrix in regiters.
```
```suggestion
Specifying the usage of the matrix: matrix left (A), matrix right (B) or accumulator (C) is required by backend implementations to reason about the layout of the matrix in registers.
```

Also (C) is rendered as the copyright sign so this needs to be corrected somehow.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+enum class use {
+  a,
+  b,
+  accumulator,
```
```suggestion
  accumulator
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+
+
+#### Layout
+Besides row major and column major layouts, `layout` is flexible enough to introduce custom layouts such as symmetric or tiled layouts.
```
It isn't clear here as a reader how "symmetric" or "tiled" layouts relate to the values of the enum class "layout" below.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,647 @@
+}
+```
+
+IMPORTANT: In both AMX and DPAS support, layout template parameter is unused and will be ignored if specified
````
```suggestion
IMPORTANT: In both AMX and DPAS support, the layout template parameter is unused and will be ignored if specified
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,647 @@
+}
+```
+
+IMPORTANT: In both AMX and DPAS support, `use` template parameter is required
````
```suggestion
IMPORTANT: In both AMX and DPAS support, the `use` template parameter is required
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+and performance, and **future versions of this API may change in ways that are incompatible with this experimental version**. The current implementation provides support of the matrix interface on Intel(R) Advanced Matrix Extensions (AMX) and DPAS_
+
+## Introduction
+This document presents an ongoing work towards defining a unified matrix interface. This interface is intended to unify different tensor hardware: Intel AMX in CPUs, DPAS in Intel GPUs, Habana Gaudi and Goya tensor and gemm cores, Nvidia TPUs, IBM Power MMA. All these hardware provide low-level intrinsics or assembly to access and perform matrix operations. The goal is to provide a unified interf ...
```
```suggestion
This document presents an ongoing work towards defining a unified matrix interface. This interface is intended to unify different matrix hardware: Intel AMX in CPUs, DPAS in Intel GPUs, Habana Gaudi and Goya tensor and gemm cores, Nvidia TPUs, IBM Power MMA. All these hardware provide low-level intrinsics or assembly to access and perform matrix operations. The goal is to provide a unified interface that is portable but also benefit from the maximum performance these different hardware can offer.
```

I personally think it is better to avoid using "tensor" apart from where absolutely necessary :i.e for Tensor Cores (c), because the word "tensor" means something different from "matrix": a matrix is a tensor if there exists transformational rules for the matrix elements with respect to a coordinate system. All (any-dimensional) matrices are not tensors although a matrix can also be a tensor: In situations where all matrices operated on are either all tensors or all not tensors it doesn't make much difference whether we refer to "matrices" as "tensors", but for applications where both "matrices" and "tensors" are used, proper use of language becomes important.

I'm not sure where the trend of replacing (any-dimensional)"matrix" with "tensor" originated, perhaps it was when "tensorflow" sounded better than "matrixflow", but I don't think there is really a reason for adopting this trend in this extension, particularly since we have adopted the naming conventions of `joint_matrix` etc rather than e.g. `joint_tensor`.

dkhaldi replied:
A matrix is technically a 2d tensor and they are both mathematical objects.
tensorflow supports more than 2 dimensions. That's why I believe it is not called matrixflow.

I tend to refer to the hardware as tensor hardware because even though they are matrix hardware (supports only 2 dimensions), they enable ndimensional tensor computing (if the code is inside loops and such). Also, they are called TPUs in general in the literature. That's why the Nvidia tensor cores are called as such for instance. Habana uses tensor cores terminology as well.

Of course, we should refer to this extension as matrix because it enables only 2d computations.

JackAKirk replied:
A matrix is only a 2-tensor if it has a basis independent meaning, likewise for a 3d matrix (3-array) etc. I personally think that Tensor Processing Unit is a terrible name for the literature to have adopted for Processing Units that perform basis dependent computations on 2-dimensional arrays (matrices) that may be part of n-dimensional arrays (n-dimensional matrices), that may or may not be part of a program that grants the matrix a tensorial nature. Here is an article that mentions the question of over-use of tensor in ML, https://www.linkedin.com/pulse/tensorsand-applications-ml-demystified-andriy-strepetilov/, with a mostly reasonably description of the meaning of tensor. It's not worth worrying about too much but I guess... for a ML algorithm you can probably define a conservation law (conserving the output) and then define a geometry (set of transformations that satisfy the conservation law) for the inputs such that the inputs are invariant with respect to those transformations (then they are tensors). So maybe "tensor"flow is justified that way.. But for the TPUs I don't see how using "tensor" leads to anything but confusion personally. If we adopt "tensor" to mean n-matrix (n-array) then what will we call something that has tensorial properties... 
Probably it doesn't matter at this point because as you say TPU appears to have become accepted as the general term in the literature. But when a word is misappropriated in society, meaning tends to be lost, and society can become poorer for it.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+
+IMPORTANT: The WI data to joint matrix mapping coordinates information is not implemented yet.
+
+IMPORTANT: Since the current tensorcores implementation is AOT, it is possible to know how many elements are owned by each WI at compile time. In this case, `wi_data` can be of type `marray`. An additional interface will be provided for the tensorcores AOT backend. 
```
```suggestion
IMPORTANT: Since the current Tensor Cores  implementation is AOT, it is possible to know how many elements are owned by each WI at compile time. In this case, `wi_data` can be of type `marray`. An additional interface will be provided for the Tensor Cores AOT backend. 
```

Actually there are some aspects of the Tensor Cores backend that are not AOT, since there is JIT compilation from the ptx level to SASS level when the code is run. I think this would be more accurate and simpler personally:

```suggestion
IMPORTANT: It is possible to know how many elements are owned by each WI at compile time. In this case, `wi_data` can be of type `marray`. An additional interface will be provided for the Tensor Cores backend. 
```

dkhaldi replied:
AOT is used here to reflect the SYCL compilation flow: If it does not use SPIRV and target is specified at compile time so this is basically AOT compilation for SYCL. But I will change the sentence to the one that does not talk about the AOT detail.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+
+
+## VNNI/Packed Layout
+Intel AMX and DPAS compute assumes register for B tile (src1) to be in VNNI format as they need 32bit of K-data in A and B to be contiguous in memory.
```
```suggestion
Intel AMX and DPAS compute assumes that the B tile register (src1) is in the VNNI format as they need 32bit of K-data in A and B to be contiguous in memory.
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+range<2> L = {1, SG_SIZE};
+int8_t *memA = malloc_shared<int8_t>(M*K, q);
+int8_t *memB = malloc_shared<int8_t>(K*N, q);
+Int32_t *memC = malloc_shared<int32_t>(M*N, q);
```
```suggestion
int32_t *memC = malloc_shared<int32_t>(M*N, q);
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+
+- Default values: this provides a default shape if the user does not provide a specific combination. In this case, aliases to the `joint_matrix` type can be used, namely `joint_matrix_a/b/c` where no additional argument is needed. This form happens when the user specifies all template parameters except the sizes of the matrices (`tiles`) M, N, and K.
+
+- General query: the general query interface provides information  about sizes, types, static/dynamic, and scopes that are supported by a specific TPU implementation. This is needed to avoid padding by the user, for tuning, and efficient code generation if used by a library. The general query return an array of `combinations` of `combination` type. Each combination includes the sizes and the type ...
```
```suggestion
- General query: the general query interface provides information  about sizes, types, static/dynamic, and scopes that are supported by a specific TPU implementation. This is needed to avoid padding by the user, for tuning, and efficient code generation if used by a library. The general query returns an array of `combinations` of `combination` type. Each combination includes the sizes and the types for the matrices A, B, and C. Note that for each TPU, the query returns `max_msize, max_nsize, max_ksize` or `msize, nsize, ksize` exclusively, depending on whether the implementation supports a continuous or discrete number of sizes. For example, the Intel AMX implementation supports a continuous number of sizes, so the `max_*` variant is applied and only the maximum number is returned. The DPAS implementation, on the other hand, supports a discrete list of numbers so the `msize, nsize, ksize` variant is applied.  This form takes place when users only specify the TPU they are interested in using.
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+|`defaultK`|  validation, default values|when no sizes are provided by the user, indicates the suggested default size for K; usually this corresponds to the maximum size the implementation supports. In validation mode, where the user does provide sizes, this is the same value K that the user provides if K is supported by the implementation
+|`joint_matrix_a`|  validation, default values|type alias for `joint_matrix` for matrix A
+|`joint_matrix_b`| validation, default values| type alias for `joint_matrix` for matrix B
+|`joint_matrix_c`|  validation, default values| type alias for `joint_matrix` for matrix C
```
Should this refer to `joint_matrix_accumulator`?

dkhaldi replied:
This is a good question. What do you think @gmlueck ?

gmlueck replied:
You have this inconsistency throughout the spec.  In many places, you refer to the three matrices as `a`, `b`, and `c`.  However, the `use` enum refers to them as `a`, `b`, and `accumulator`.  You should decide on one or the other and then use it consistently throughout.  Since you call the first two matrices `a` and `b`, it seems natural to me to call the third matrix `c`.  Is there a good reason to call the third matrix `accumulator`?

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+| Member/type alias in `tpu_params` | Forms they are defined in |Description
+|`type_a`| validation, default values|type alias for the type of matrix A
+|`type_b`|  validation, default values|type alias for the type of matrix B
+|`type_c`|  validation, default values|type alias for the type of matrix C
```
Should this be `type_accumulator` for consistency with `matrix::use`?

dkhaldi replied:
This is a good question. @gmlueck, what do you think?

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+We did not utilize this extension for this matrix API version because sub-group local memory is not yet well defined in {dpcpp}. Moreover, the representation of this notion in LLVM IR and SPIR-V is not clear yet. 
+
+### WI data to joint matrix mapping coordinates information for piece-wise operations
+The indexing provided inside the `wi_data` class acesses only the portion of the current WI. It is not possible the location or coordinates of this portion in the original matrix.  This coordinates mapping  is implementation defined and change from one backend to the other.   For general piece-wise operations like sum of rows of a matrix, the WI data to joint matrix mapping coordinates informatio ...
```
```suggestion
The indexing provided inside the `wi_data` class accesses only the portion of the matrix held by the current WI. It is not possible to know the location of this portion in the original matrix.  This coordinates mapping  is implementation defined and changes from one backend to the other.   For general piece-wise operations like sum of rows of a matrix, the WI data to joint matrix mapping information is needed to reason about the matrix view.
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+
+### WI data to joint matrix mapping coordinates information for piece-wise operations
+The indexing provided inside the `wi_data` class acesses only the portion of the current WI. It is not possible the location or coordinates of this portion in the original matrix.  This coordinates mapping  is implementation defined and change from one backend to the other.   For general piece-wise operations like sum of rows of a matrix, the WI data to joint matrix mapping coordinates informatio ...
+With joint matrix, we want to write, as much as possible, one code to run on different backends. So if backend X states that a WI owns one exact row of the matrix for instance. Writing the following code will work only on that backend for that version of hardware. The hardware and implementations change, for instance, the same WI can own half of the row because SG size increased or hardware units ...
```
```suggestion
Within the joint matrix extension, we want to write, as much as possible, one code to run on different backends. If backend X states that a WI owns one exact row of the matrix for instance, writing the following code will work only on that backend for that version of hardware. If a different hardware and implementations is used, the same WI may own half of the row if the SG size decreased for example. 
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+
+
+
+We want to keep backward compatibility in the joint matrix code when implementations or hardware change. To that end, instead of hard-code this mapping, we write  general backend and target-agnostic, especially in the JIT compilation mode of SYCL. This is possible by querying this mapping so code does not have to change from one version to the other.
```
```suggestion
We want to keep backward compatibility in the joint matrix code when implementations or hardware change. To that end, instead of hard-coding this mapping, we use general backend and target-agnostic functionality, especially in the JIT compilation mode of SYCL. This is possible by querying this mapping so code does not have to change from one version to the other.
```

I find the last sentence confusing "This is possible by querying this mapping so code does not have to change from one version to the other"

dkhaldi replied:
sounds good

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,647 @@
+- In the future looking APIs, `get_wi_data` (that is currently under design) returns an owned object. Should this return a view object to make sure the original matrix C is changed after its slices are modified.
+
+## TODO List
+- Add WI data to joint matrix mapping coordinates information for piece-wise operations. This will be added as part of the query or new methods to the 'get_wi_data' class. 
```
```suggestion
```

dkhaldi replied:
This is still an outstanding task. Why do you want to remove it?

JackAKirk replied:
Sorry I misread the task. Fine to keep it.

dkhaldi commented:
> I have a question about the old specification, which is now named "sycl_ext_oneapi_deprecated_matrix_no_use". Is that version of the API still supported? 

@gmlueck, yes, this version is still supported and will be supported until the new API is implemented on the GPU. Currently, it is only implemented on the CPU.  

> Does (or will) the implementation print a warning message for applications that use the old API?

No it does not because the new API is not supported on the GPU yet. So for users, they still need to use the old API until we remove it for oneAPI 2023.1 release.
> 
> Or, is the old API no longer supported, and it is (or will be) removed?
The old API will be removed for oneAPI 2023.1 release.

> Both options are OK for an experimental extension like this. However, we should treat the documentation differently in the two cases.

For this release, we want to keep the old API unchanged and supported. At the same time, we move towards the new API. Once the new API is also supported on the GPU, we can completely remove the old API.

JackAKirk reviewed (approved):
LGTM

dkhaldi commented:
@intel/dpcpp-specification-reviewers, any more review on this?

gmlueck commented:
> For this release, we want to keep the old API unchanged and supported. At the same time, we move towards the new API. Once the new API is also supported on the GPU, we can completely remove the old API.

Sorry for the delayed response.  It looks like both extensions define the same types in different ways.  For example, the new version has:

```
namespace sycl::ext::oneapi::experimental::matrix {
template <typename T, use Use,
          size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent,
          layout Layout = layout::unused, typename Group = sub_group>
struct joint_matrix {
    joint_matrix(Group g) {}
};
}
```

and the old version has:

```
namespace sycl::ext::oneapi::experimental::matrix {
template <typename T, size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent,
          matrix_layout Layout = matrix_layout::row_major, typename Group = sub_group>
struct joint_matrix {
    joint_matrix(Group g) {}
};
}
```

How can you support both simultaneously?  How does the user determine whether they get the new vs. the old API?

dkhaldi commented:
> How can you support both simultaneously? How does the user determine whether they get the new vs. the old API?

they can do this by specifying the SYCL_EXT_ONEAPI_MATRIX feature test macro. See what we have today here: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/matrix/matrix.hpp

gmlueck commented:
> they can do this by specifying the SYCL_EXT_ONEAPI_MATRIX feature test macro. See what we have today here: https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/matrix/matrix.hpp

This is not how the feature-test macros are supposed to work.  The *implementation* determines the value of the feature-test macro, not the application.  (The extension specification says this in the section "Feature test macro".)  If you want the user to choose the API via a macro, we should choose a different macro name.  For example, `SYCL_EXT_ONEAPI_MATRIX_LEGACY_API` could enable the old API.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,644 @@
+|======================
+|Value |Description
+|1     |Initial extension JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported 
+|2     |joint matrix type has a new `use` parameter. `layout` on matrix is unused. JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported 
```
The `Layout` template parameter in this version seems confusing.  It seems like it is required to be `layout::unused` for most of the functions.  For example, `joint_matrix_load`, `joint_matrix_store`, and `joint_matrix_mad` all require the layout to be `layout::unused`.  Why not just remove that template parameter?  It seems like you can't set it to any other value.

dkhaldi replied:
We cannot remove it because Jack requires it for the CUDA backend on the mad function. But, you are right, it should not be set to unused for these functions. This was based on a comment from Jack but I might have misunderstood him.  
@JackAKirk,  can you please comment here on how you will be able to set it for other values if it is unused?

JackAKirk replied:
Yes on the mad function it is required for the cuda backend. If we want to have a single `joint_matrix_mad` function for all backends I suggested previously that we can however use the template parameter with a default value that is set layout::unused for the intel backend: I think this allows as far as the programmer is concerned the same interface as you would have if you delete the layout template parameter: as I've and implemented here: https://github.com/intel/llvm/pull/6657/files#diff-f71a436bdeda598b29caad471fa637a2844a12f38fe4e85b15b2ccb37bd09833R825

For the `joint_matrix_load` `joint_matrix_store` the CUDA backend uses two overloads of these functions: one for the c/accumulator matrix which matches exactly to the intel backend interface of these functions, and another overload that requires the template parameter of joint layout and doesnt have the regular parameter of joint layout: For the intel backend you don't need to worry about the second overload, but again, I think it would be nice if the first overload in the CUDA backend coincides with the Intel backend, so that only one function is used for both backends.

JackAKirk replied:
The interfaces in this spec doc for `joint_matrix_load` and `joint_matrix_store` are already fine from my point of view. As described above there will can be an additional overload of `joint_matrix_load` wrapped in `#ifdef nvptx` that deals with the loads of use::a/use::b where the layout is required as a template parameter in the CUDA backend.

JackAKirk replied:
> We cannot remove it because Jack requires it for the CUDA backend on the mad function. But, you are right, it should not be set to unused for these functions. This was based on a comment from Jack but I might have misunderstood him. @JackAKirk, can you please comment here on how you will be able to set it for other values if it is unused?

In Summary I think the Layout parameter should be set to unused by default in `joint_matrix_mad`, but it is not necessary at all in the other functions (which is actually as the current spec doc describes).

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,644 @@
+namespace sycl::ext::oneapi::experimental::matrix {
+template <typename T, use Use,
+          size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent,
+          layout Layout = layout::unused, typename Group = sub_group>
```
You should be consistent with the spelling of the template parameters in this specification.  Sometimes you use `Layout` and `Use` but in other places you use `L` and `U`.  Choose one or the other.  Personally, I prefer the spelled out versions `Layout` and `Use`.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,644 @@
+}
+```
+
+IMPORTANT: In both AMX and DPAS support, the `layout` template parameter is unused and will be ignored if specified
````
This does not match the code synopses you have below, which require they `layout` template parameter to be `layout::unused`.

dkhaldi replied:
are you referring to this code?
```c++
joint_matrix<int8_t, use::a, tM, tN> tA(sg);
```
yes this code is considering layout to be "unused".
What this statement is saying is that:  in joint matrix, layout is optional. By default is it "unused". In both AMX and DPAS support, layout is unused. If it is specified for tensorcores support (where it is required. Jack will add documentation on top of this once this PR is merged), it is ignored in AMX and DPAS implementations. 
Does this make sense?

gmlueck replied:
I'm referring to code synopsis like this:

```
template <typename Group, typename T, size_t NumRows, size_t NumCols,
          use Use, access::address_space Space>
void joint_matrix_load(Group sg,
  joint_matrix<T, Use, NumRows, NumCols, layout::unused, Group> &res,
  multi_ptr<T, Space> src, size_t stride, layout memL);
```

That synopsis says that the matrix parameter must be `layout::unused`.  Any other layout value will cause a compilation error.  Therefore, the value is not ignored.  The same situation exists for the synopses of `joint_matrix_store` and `joint_matrix_mad`.

dkhaldi replied:
You are right, I am adding Layout on these three functions as layout can have a different value for the CUDA backend.

gmlueck reviewed (commented):
Whenever you rename (or move) a specification, you should search the repo and see if there are any links to the old name.  In your case, there are several references from "sycl/ReleaseNotes.md".  These should be updated:

* Point to the old file, or
* Point to the new file, or
* Change the wording to remove the link.

dkhaldi commented:
>> You have this inconsistency throughout the spec. In many places, you refer to the three matrices as a, b, and c. However, the use enum refers to them as a, b, and accumulator. You should decide on one or the other and then use it consistently throughout. Since you call the first two matrices a and b, it seems natural to me to call the third matrix c. Is there a good reason to call the third matrix accumulator?

@JackAKirk what do you think?

dkhaldi commented:
> Whenever you rename (or move) a specification, you should search the repo and see if there are any links to the old name. In your case, there are several references from "sycl/ReleaseNotes.md". These should be updated:
> 
> * Point to the old file, or
> * Point to the new file, or
> * Change the wording to remove the link.

Thank you! 
yes, I will point to the new file

JackAKirk commented:
> > > You have this inconsistency throughout the spec. In many places, you refer to the three matrices as a, b, and c. However, the use enum refers to them as a, b, and accumulator. You should decide on one or the other and then use it consistently throughout. Since you call the first two matrices a and b, it seems natural to me to call the third matrix c. Is there a good reason to call the third matrix accumulator?
> 
> @JackAKirk what do you think?

I think that the argument for having a,b and accumulator is that the programming idiom is 
```
d=a*b+c
```
Where d would have to be use::c(use::accumulator)
or  as in most use cases:
```
c=a*b+c
```
`use::accumulator` is a more general name than `use::c`. However I think either choice is fine.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,644 @@
+}
+```
+
+IMPORTANT: In both AMX and DPAS support, the `layout` template parameter is unused and will be ignored if specified
````
If this document will focus on the matrix extension interfaces apparent to the user for the Intel backend only, you could consider completely omitting this "Important" statement, and describe joint_matrix as above without the template parameter as you did with:
```
joint_matrix<int8_t, use::a, tM, tN> tA(sg);
```

However admittedly this could be perhaps considered a little dodgy since the user still could construct a joint_matrix with the layout template parameter so long as it is layout::unused, assuming that we keep it as it is.
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,644 @@
+
+Note that in order to get maximum performance on Intel AMX and DPAS, prepacking data in the memory is necessary. If users did not specify the packed layouts, transforms done by the implementation will be slow due to extra scatter/gather operations. Hence, we expose the `packed` layout to the user to specify that A or B have already been VNNIed. The packed or VNNI layout is introduced in the `VNNI ...
+
+IMPORTANT: In the current AMX and DPAS implementation, the layout in the load of matrix B must be `packed` or `row_major`. Automatic VNNI transform is supported. The layout in the load of matrices A and C must be `row_major`, and the layout in the store of matrix C must also be `row_major`.
```
Could consider being more precise with what you mean by "the layout" here, by referencing the defs of joint_matrix_load/store. My suggestion is very rough and can be structured better.

```suggestion
IMPORTANT: In the current AMX and DPAS implementation, the layout in the load of matrix B (provided by the `layout memL` parameter below) must be `packed` or `row_major`. Automatic VNNI transform is supported. The layout in the load of matrices A and C must be `row_major`, and the layout in the store of matrix C (provided by the `layout memL` parameter below) must also be `row_major`.
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -150,9 +150,9 @@ This function loads data from memory to the 2d tiles/registers of Intel AMX/DPAS
 namespace sycl::ext::oneapi::experimental::matrix {
   template <typename Group, typename T, size_t NumRows, size_t NumCols,
-          use Use, access::address_space Space>
+          use Use, layout Layout, access::address_space Space>
```
Why is the layout template parameter added back?
Even in the CUDA backend this is never necessary for joint_matrix_store.

dkhaldi replied:
as you can see, I did not add for C on the mad function.
But for store, remember in Intel backend, you can store into A or B. So if the code has A with layout other than unused, store will not fail.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,659 @@
+
+== Notice
+
+Copyright (c) 2021-2021 Intel Corporation.  All rights reserved.
```
```suggestion
Copyright (c) 2021-2022 Intel Corporation.  All rights reserved.
```
maybe?

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,659 @@
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+This extension is written against the SYCL 2020 revision 3 specification.  All
```
```suggestion
This extension is written against the SYCL 2020 revision 5 specification.  All
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,659 @@
+
+**_NOTE:_** _This document describes the current design and API for the matrix
+extension to {dpcpp}. This is an initial experimental version to try out functionality
+and performance, and **future versions of this API may change in ways that are incompatible with this experimental version**. The current implementation provides support of the matrix interface on Intel(R) Advanced Matrix Extensions (AMX) and DPAS_
```
```suggestion
and performance, and **future versions of this API may change in ways that are incompatible with this experimental version**. The current implementation provides support of the matrix interface on Intel(R) Advanced Matrix Extensions (AMX), DPAS and Nvidia(R) Tensor Cores_
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,659 @@
+and performance, and **future versions of this API may change in ways that are incompatible with this experimental version**. The current implementation provides support of the matrix interface on Intel(R) Advanced Matrix Extensions (AMX) and DPAS_
+
+## Introduction
+This document presents an ongoing work towards defining a unified matrix interface. This interface is intended to unify different tensor hardware: Intel AMX in CPUs, DPAS in Intel GPUs, Habana Gaudi and Goya tensor and gemm cores, Nvidia TPUs, IBM Power MMA. All these hardware provide low-level intrinsics or assembly to access and perform matrix operations. The goal is to provide a unified interf ...
```
Should we mention hardware that isn't supported? I think that could be confusing. I've removed mentions of hardware that isn't currently supported like goya etc.
```suggestion
This document presents an ongoing work towards defining a unified matrix interface. This interface is intended to unify different tensor hardware that provide low-level intrinsics or assembly to access and perform matrix operations. The goal is to provide a unified interface that is portable but also benefit from the maximum performance these different hardware can offer.
```

dkhaldi replied:
This is showing the goal of the API. I don't see a problem with keeping the other hardware. The overall goal will encourage external support on the unsupported hardware.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,659 @@
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+template <typename T, use Use,
+          size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent,
````
Since `sycl::dynamic_extent` extent isn't going to be immediately implemented should we avoid mentioning it here?
```suggestion
          size_t Rows, size_t Cols,
```

dkhaldi replied:
This is a big subject and will need a lot of discussion. 
I would defer this to once this is proposed to SYCL.
But I will add a note about this

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,659 @@
+  template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          use Use, access::address_space Space>
+  void joint_matrix_load(Group sg,
+    joint_matrix<T, Use, NumRows, NumCols, layout::dynamic, Group> &res,
```
```suggestion
    joint_matrix<T, use::accumulator, NumRows, NumCols, layout::dynamic, Group> &res,
```

dkhaldi replied:
see above

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,659 @@
+
+IMPORTANT: The WI data to joint matrix mapping coordinates information is not implemented yet.
+
+IMPORTANT: In the Tensore Cores implementation, it is possible to know how many elements are owned by each WI at compile time. In this case, `wi_data` can be of type `marray`. An additional interface will be provided for the Tensor Cores backend.
```
```suggestion
IMPORTANT: In the Tensor Cores implementation, it is possible to know how many elements are owned by each WI at compile time. In this case, `wi_data` can be of type `marray`. An additional interface will be provided for the Tensor Cores backend.
```

JackAKirk commented:
LGTM. I just suggested some small changes.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,659 @@
+```c++
+namespace sycl::ext::oneapi::experimental::matrix {
+  template <typename Group, typename T, size_t NumRows, size_t NumCols,
+          use Use, access::address_space Space>
````
Currently, in the Intel backend, load to joint matrix that has layout::dynamic can be done for matrices A and B as well.
so this is not just for A and B.
One next step will be to get rid of these divergences (See note IMPORTANT: In both AMX and DPAS support, the `layout` template parameter can be `dynamic` for Matrix A and B as well.). At that time, this change you are proposing will make full sense if we decide to move the Intel specific case to a different extension.

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,661 @@
+namespace sycl::ext::oneapi::experimental::matrix {
+  template <typename Group, typename Ta, typename Tb, typename Tc, std::size_t M, std::size_t K, std::size_t N, 
+            layout LayoutA, layout LayoutB>
+  joint_matrix<Group, use::accumulator, Td, M, N, layout::dynamic, Group> joint_matrix_mad(Group sg,
```
```suggestion
  joint_matrix<Tc, use::accumulator, M, N, layout::dynamic, Group> joint_matrix_mad(Group sg,
```

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,661 @@
+  template <typename Group, typename Ta, typename Tb, typename Tc, std::size_t M, std::size_t K, std::size_t N, 
+            layout LayoutA, layout LayoutB>
+  joint_matrix<Group, use::accumulator, Td, M, N, layout::dynamic, Group> joint_matrix_mad(Group sg,
+    joint_matrix<Ta, use::a, M, K, layoutA, Group> A,
```
```suggestion
    joint_matrix<Ta, use::a, M, K, layoutA, Group> &A,
```

etc

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,661 @@
+    joint_matrix<T, Use, NumRows, NumCols, layout::dynamic, Group> &res,
+    multi_ptr<T, Space> src, size_t stride, layout memL);
+    
+  template <typename Group, typename T, size_t NumRows, size_t NumCols,
```
```suggestion
  template <typename Group, typename S, typename T, size_t NumRows, size_t NumCols,
          use Use, layout Layout, access::address_space Space>
  void joint_matrix_load(Group sg,
    joint_matrix<S, Use, NumRows, NumCols, Layout, Group> &res,
    multi_ptr<T, Space> src, size_t stride);
```

We need typename S as an implementation detail so we can support const T cases via: std::enable_if_t<std::is_same<S, std::remove_const_t<T>>::value

dkhaldi commented:
@intel/dpcpp-doc-reviewers, any more reviews on this?
It will be good if we can merge this ASAP.

dkhaldi commented:
@intel/dpcpp-doc-reviewers, can you please help merge this if there are no more comments?

pvchupin commented:
@dkhaldi, please move deprecated extension to https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/deprecated

dkhaldi commented:
> @dkhaldi, please move deprecated extension to https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/deprecated

done

dkhaldi commented:
@intel/dpcpp-doc-reviewers, please help merge.

pvchupin commented:
Ping @intel/dpcpp-specification-reviewers

gmlueck commented:
> https://github.com/orgs/intel/teams/dpcpp-doc-reviewers, please help merge.

Sorry for the delay.  This is the next thing on my review list.

gmlueck commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,661 @@
+|Value |Description
+|1     |Initial extension JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported. The exact API used for this implementation is detailed in [matrix extension](doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_deprecated_matrix_no_use.asciidoc)
+|2     |joint matrix type has a new `use` parameter. `layout` on matrix is optional for `C` matrix. JIT implementation on Intel AMX and DPAS. load, store, mad, fill, piece-wise operations, and the query interface are supported 
+|3     |Implementation on Nvidia Tensor Cores
```
Does this specification also apply to Nvidia, or is that still a separate specification?  If it is still separate, can there be a link here to that other spec?

JackAKirk replied:
Here is the PR for the sycl_ext_oneapi_matrix_cuda extension that is fully compatible with this extension (sycl_ext_oneapi_matrix) but built on top of it as an optional (cuda only extension) https://github.com/intel/llvm/pull/6968

dkhaldi replied:
This specification does apply to CUDA. I expect Jack to add some of the restrictions/details that are missing once this is merged. #6968 contains the additional functions that only applies to CUDA backend. @JackAKirk, the file you add in #6968 should be included in the new matrix folder, right?

JackAKirk replied:
> This specification does apply to CUDA. I expect Jack to add some of the restrictions/details that are missing once this is merged. #6968 contains the additional functions that only applies to CUDA backend. @JackAKirk, the file you add in #6968 should be included in the new matrix folder, right?

Yeah sure I can move it to that folder.

gmlueck reviewed (commented):
Some questions below, but also one general comment:

Since you moved the deprecated version of the API spec to "doc/extensions/deprecated", there isn't any need for the "doc/extensions/experimental/sycl_ext_oneapi_matrix" directory.  Can you move the spec from:

sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc

to:

sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix.asciidoc

This also means that you can undo the changes to "ReleaseNotes.md".

JackAKirk commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,661 @@
+namespace sycl::ext::oneapi::experimental::matrix {
+template <typename T, use Use,
+          size_t Rows=sycl::dynamic_extent, size_t Cols=sycl::dynamic_extent,
+          layout Layout = layout::dynamic, typename Group = sub_group>
```
> Can you remind me of the rules for using `Layout` if you want portable code that works on both Intel and CUDA? I thought the "A" and "B" matrices couldn't be dynamic if you wanted to run on CUDA, is that right?
> 

Yes the A/B matrices cannot be dynamic for cuda: we use partial specializations for all valid joint_matrix declarations in the cuda backend, so it isn't possible to construct such a use::a/use::b joint_matrix with layout::dynamic in the cuda backend.
> We want the default template parameters to result in portable code, and I see that the template parameter here defaults to `dynamic`. Is that portable?
> 
> Don't we also want a description somewhere in this spec explaining how to use "Layout" if you want portable code? I don't see anything like that now.

FYI for the cuda backend it is also useful for layout::dynamic to be set as default because it means users don't even need to provide a layout template argument when constructing an accumulator joint_matrix: As here in the test for the portable interfaces:

https://github.com/intel/llvm-test-suite/blob/b8a6c135726072ab3e8ab07cb73b2c1d81fbafb0/SYCL/Matrix/joint_matrix_tensorcores.cpp#L151

gmlueck replied:
> Yes the A/B matrices cannot be dynamic for cuda: we use partial specializations for all valid joint_matrix declarations in the cuda backend, so it isn't possible to construct such a use::a/use::b joint_matrix with layout::dynamic in the cuda backend.

This restriction isn't conveyed by this spec.  Do we expect this specification to describe a common API that includes CUDA?

> FYI for the cuda backend it is also useful for layout::dynamic to be set as default because it means users don't even need to provide a layout template argument when constructing an accumulator joint_matrix

I agree that this is useful.  This could be achieved by defaulting the `Layout` parameter conditionally, depending on the value of `Use`.

dkhaldi replied:
Please see under "Layout" subsection:
Important | In both AMX and DPAS support, the layout template parameter can be dynamic for Matrix A and B as well.

I only talk about what is allowed for AMX and DPAS. This is not conveying the restriction about the CUDA backend because I expect Jack to update this file after it is merged to  add the CUDA backend restrictions in joint matrix type (dynamic is the default for only matrix C) and in the joint_matrix_load (load that takes layout as argument is only for C matrix). 

So the restrictions that apply for the CUDA backend will be explained in the specification text by Jack and not imposed in the syntax. If imposed in the syntax, this means we are restricting Intel backend as well.

@JackAKirk, @gmlueck  do you want to proceed differently?

gmlueck replied:
> So the restrictions that apply for the CUDA backend will be explained in the specification text by Jack and not imposed in the syntax. If imposed in the syntax, this means we are restricting Intel backend as well.

I thought one of the outcomes of our meeting a few weeks ago was to prioritize the portable case.  Therefore, I think we want the default values of the template parameters to work for both Intel and CUDA.  My concern above is that `Layout` defaults to `dynamic`, but that doesn't work on CUDA for the "A" or "B" matrices.

dkhaldi replied:
You are right about prioritizing the portable case.  But one of the outcomes of that meeting as well is that the restrictions will still be made in the text of the spec, not in the syntax. Not just yet.

If we make the restriction in the syntax, we are excluding the Intel flexibility.
In the text, we should add that for the CUDA backend, dynamic as default applied to only matrix accumulator.

One of the reasons of that outcome is that I did not want to make more changes at a time. 
The change you are suggesting (in the syntax) should be made in a separate PR after we discuss that specific issue more.

gmlueck replied:
Can you please add this as an open issue in the list at the bottom of this spec, then?  I just want to make sure we do not forget about issues like this.

JackAKirk replied:
Here is a full implementation of the agreed interfaces: https://github.com/intel/llvm/pull/7077 with a working cuda implementation. See that we still use the layout::dynamic default parameter here: https://github.com/intel/llvm/blob/36004a041513635d2567a7d77d9ebab38ba3d29c/sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcores.hpp#L33
but as I mentioned the cuda implementation won't allow a joint_matrix to be constructed with (use::a || use::b) && layout::dynamic.

dkhaldi replied:
I will add it to open issues.
Note that this is not the only non portable code that is included in this document. We still need to sort things like "packed" layout. 
Once we have the full support of this new API (in a month or so) mentioned by Jack (#7077). I will open up discussion for the non portable scenarios that are still in this document.

JackAKirk replied:
> Please see under "Layout" subsection: "Important | In both AMX and DPAS support, the layout template parameter can be dynamic for Matrix A and B as well."
> 
> I only talk about what is allowed for AMX and DPAS. This is not conveying the restriction about the CUDA backend because I expect Jack to update this file after it is merged to add the CUDA backend restrictions in joint matrix type (dynamic is the default for only matrix C) and in the joint_matrix_load (load that takes layout as argument is only for C matrix).
> 
> So the restrictions that apply for the CUDA backend will be explained in the specification text by Jack and not imposed in the syntax. If imposed in the syntax, this means we are restricting Intel backend as well.
> 
> @JackAKirk, @gmlueck do you want to proceed differently?

I think this is fine.
If you think that some information is more appropriate in a cuda only extension or intel only extension or the main matrix extension, then I don't mind moving it, especially since these extensions are all experimental. 

I imagine that it is more traditional that extensions are additive rather that subtractive, but I don't think this is a big issue at the moment: so long as the interfaces (note that I don't say "code" here since DPAS and Tensor Cores don't ever have cases where parameters provided to the interfaces are portable) written by a user are fully portable across cuda and intel for the cases where the hardware allows them to (all the practically useful cases) then this is the main thing.

dkhaldi commented:
> Since you moved the deprecated version of the API spec to "doc/extensions/deprecated", there isn't any need for the "doc/extensions/experimental/sycl_ext_oneapi_matrix" directory. Can you move the spec from:
> 
> sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc
> 
> to:
> 
> sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix.asciidoc

The folder is still necessary to add CUDA backend specific API (marray-based element wise operations).
@JackAKirk please comfirm.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
```diff
@@ -0,0 +1,661 @@
+  void joint_matrix_store(Group sg,
+    joint_matrix<T, Use, NumRows, NumCols, Layout, Group> &res,
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, layout memL);
+}
```
No need for two versions. In both Intel and CUDA backends, store must take layout as a dynamic argument.
@JackAKirk please confirm

gmlueck replied:
What if the template parameter of the matrix is something other than `dynamic`, though?  For example, consider the case when `Layout` is `row_major`.  Does it still make sense to require the user to specify a layout parameter at runtime to `joint_matrix_store`?  What if the layout they specify at runtime is different from the `Layout` template parameter?  Is that OK?

dkhaldi replied:
In the CUDA case, you can only store to matrix accumulator.
For the Intel backend, Besides C, you can store to matrix A or B. In this case, if users decide to do that, they assume it is not portable code, so they would use the version of joint matrix where A and B type layouts are dynamic.

JackAKirk replied:
This is not something we needed to consider in the Nvidia backend because the nvidia backend doesn't support native implementations of `joint_matrix_store` for anything other than accumulator (and we don't plan on implementing it at all). My understanding is that the intel backend may want to support `joint_matrix_store` for use::a and use::b and that it can choose whatever interface it wants for this overload of `joint_matrix_store`. I thought that this would eventually be added to an intel specific extension document, similar to what we did for cuda in https://github.com/intel/llvm/pull/6968.

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,663 @@
+}
+```
+
+IMPORTANT: Matrix layout defaulting to `layout::dynamic` applies only to matrix with `use::accumulator`
````
There is no default for use::a and use::b. the default value `dynamic` applies only to matrix::accumulator. 
In the commit comment, by "restrict the API in this document to only the portable API", I meant you can only write portable code with this documented API.

Is the note in the text good enough?
"IMPORTANT: Matrix layout defaulting to `layout::dynamic` applies only to matrix with `use::accumulator`"

gmlueck replied:
> In the commit comment, by "restrict the API in this document to only the portable API", I meant you can only write portable code with this documented API.

If you still want to support `layout::dynamic` for `use::a` and `use::b`, then you have a bug in the definition of `joint_matrix_load` because that function only allows loading a matrix with dynamic layout for `use::accumulator`.

dkhaldi replied:
It is the portable API, so it should not allow `layout::dynamic` for `use::a` and `use::b`.
This case will be added to the intel-specific extension

dkhaldi commented on sycl/doc/extensions/experimental/sycl_ext_oneapi_matrix/sycl_ext_oneapi_matrix.asciidoc:
````diff
@@ -0,0 +1,663 @@
+    multi_ptr<T, Space, IsDecorated> src, size_t stride, layout memL);
+}
+```
+This function stores the data from the 2d tiles back to memory.
````
Okay

dkhaldi commented:
@intel/dpcpp-specification-reviewers, please help merge
